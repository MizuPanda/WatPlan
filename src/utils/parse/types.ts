export interface CourseRef {
    subject:   string;   // e.g. "CS"
    catalog:   string;   // e.g. "138" or "246E"
    minGrade?: number;   // e.g. 60 if “≥ 60% in CS 138”
  }
  
  export type Logic = 'AND' | 'OR' | 'N_OF';
  
  export interface AST {
    op: Logic;
    n?: number;
    nodes: (AST | CourseRef)[];
  }
  
  export interface Level { year: number; term: 'A' | 'B' }
  
  export interface Requirements {
    prereq?:        AST;
    coreq?:         AST;
    antireq?:       CourseRef[];
    minUnits?:      number;
    minLevel?:      Level;
    requiredPlans?: string[];
    bannedPlans?:   string[];
    studentGroup?:  string;
    milestones?:    string[];
    unparsedTail?:  string;
  }
  