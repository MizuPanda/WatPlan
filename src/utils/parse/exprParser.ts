// /src/parse/exprParser.ts

import { AST, CourseRef, Logic } from './types';

/** e.g. "CS138>=60" **/
const GRADE_COURSE_RE = /^([A-Z]+)(\d{3}[A-Z]?)>=(\d{1,3})$/;
/** e.g. "AFM102", "CS330E", "JAPAN201R" **/
const COURSE_RE       = /^([A-Z]+)(\d{3}[A-Z]?)$/;
/** Word→num for “One of”, etc. **/
const WORD2NUM = new Map<string,number>([
  ['one',1],['two',2],['three',3],['four',4],['five',5],
  ['six',6],['seven',7],['eight',8],['nine',9]
]);

export type Tok =
  | { kind:'COURSE'; value: CourseRef }
  | { kind:'N_OF';   n: number }
  | { kind:'AND'|'OR'|'LP'|'RP' };

/** Pulls out grade clauses into a single “CS138>=60” style token */
function encodeGradeClauses(txt:string):string {
  const gradeRE   = /(?:a\s+)?grade\s+of\s+(\d{1,3})%\s+(?:or\s+higher\s+)?in\s+([A-Za-z]+)\s+(\d{3}[A-Za-z]?)/gi;
  const atLeastRE = /at\s+least\s+(\d{1,3})%\s+in\s+([A-Za-z]+)\s+(\d{3}[A-Za-z]?)/gi;
  return txt
    .replace(gradeRE,   (_m,pct,subj,cat) => `${subj}${cat}>=${pct}`)
    .replace(atLeastRE, (_m,pct,subj,cat) => `${subj}${cat}>=${pct}`);
}

export function tokenize(rawText:string):Tok[] {
  // 1) encode grade clauses
  let text = encodeGradeClauses(rawText);

  // 2) isolate parentheses
  text = text.replace(/([\(\)])/g,' $1 ');

  // 3) collapse “SUBJ 123”→“SUBJ123”
  text = text.replace(/([A-Za-z]+)\s+(\d{3}[A-Za-z]?)/g,'$1$2');

  // 4) hyphens→spaces, split
  const words = text.replace(/-/g,' ').split(/\s+/).filter(Boolean);

  const toks:Tok[] = [];
  let lastSubj:string|null = null;

  for (let i=0; i<words.length; i++){
    let w = words[i];

    // parens
    if (w==='('){ toks.push({kind:'LP'}); continue; }
    if (w===')'){ toks.push({kind:'RP'}); continue; }

    // strip commas/semis/periods
    w = w.replace(/^[,;.]+|[,;.]+$/g,'');

    const low = w.toLowerCase();
    if (WORD2NUM.has(low) && words[i+1]?.toLowerCase()==='of') {
      toks.push({kind:'N_OF',n:WORD2NUM.get(low)!});
      i++; continue;
    }
    if (low==='and' || w===';'){ toks.push({kind:'AND'}); continue; }
    if (low==='or')           { toks.push({kind:'OR'});  continue; }

    const up = w.toUpperCase();

    // GRADED COURSE first
    const gm = GRADE_COURSE_RE.exec(up);
    if (gm) {
      const [, subj, cat, pct] = gm;
      toks.push({
        kind:'COURSE',
        value:{ subject: subj, catalog: cat, minGrade: +pct }
      });
      lastSubj = subj;
      continue;
    }

    // NORMAL COURSE
    const cm = COURSE_RE.exec(up);
    if (cm) {
      const [, subj, cat] = cm;
      toks.push({
        kind:'COURSE',
        value:{ subject: subj, catalog: cat }
      });
      lastSubj = subj;
      continue;
    }

    // BARE NUMBER → reattach lastSubj
    const dm = /^(\d{3}[A-Z]?)$/.exec(up);
    if (dm && lastSubj) {
      toks.push({
        kind:'COURSE',
        value:{ subject: lastSubj, catalog: dm[1] }
      });
      continue;
    }
  }

  return toks;
}

/** Precedence */
const PREC:Record<Logic,number> = { N_OF:3, AND:2, OR:1 };

/** Shunting‑yard → AST; wraps a single course into AND([...]) */
export function parseExpression(txt:string):AST|undefined {
  const tokens = tokenize(txt);
  const output:(AST|CourseRef)[] = [];
  const ops:Tok[] = [];

  const popOp = () => {
    const o = ops.pop()!;
    const r = output.pop()!, l = output.pop()!;
    output.push({
      op:   o.kind==='N_OF'?'N_OF': o.kind==='OR'?'OR':'AND',
      n:    o.kind==='N_OF'?(o as any).n:undefined,
      nodes:[l,r]
    });
  };

  for (const t of tokens) {
    if (t.kind==='COURSE') {
      output.push(t.value);
    }
    else if (t.kind==='N_OF' || t.kind==='AND' || t.kind==='OR') {
      const tp = PREC[t.kind==='N_OF'?'N_OF':t.kind];
      while (ops.length) {
        const top = ops[ops.length-1];
        if ((top.kind==='N_OF'||top.kind==='AND'||top.kind==='OR')
            && PREC[(top.kind==='N_OF'?'N_OF':top.kind)] >= tp) {
          popOp(); continue;
        }
        break;
      }
      ops.push(t);
    }
    else if (t.kind==='LP') {
      ops.push(t);
    }
    else {
      // RP
      while (ops.length && ops[ops.length-1].kind!=='LP') popOp();
      ops.pop();
    }
  }
  while (ops.length) popOp();

  if (output.length===1 && !('op' in output[0])) {
    return { op:'AND', nodes:[ output[0] as CourseRef ] };
  }
  return output[0] as AST | undefined;
}
