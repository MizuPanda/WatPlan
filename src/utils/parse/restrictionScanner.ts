import { Requirements } from './types';

/** split comma‑or‑and lists into array */
const planSplit = (s: string) => s.split(/,\s*| and /).map(x => x.trim());

export function scanRestrictions(tail: string, req: Requirements) {
  // break on semicolon or period to isolate clauses
  const parts = tail.split(/[;\.]/).map(p => p.trim()).filter(Boolean);

  for (const part of parts) {
    // Level at least 2A or 3B
    const lvl = /Level at least (\d)([AB])/i.exec(part);
    if (lvl) {
      req.minLevel = { year: +lvl[1], term: lvl[2] as 'A' | 'B' };
    }

    // Units
    const unit = /([0-9]+\.[0-9]{2}) unit/i.exec(part);
    if (unit) {
      req.minUnits = +unit[1];
    }

    // Milestones
    const ms = part.match(/([A-Za-z ]+? Milestone)/g);
    if (ms) {
      req.milestones = (req.milestones || []).concat(ms.map(m => m.trim()));
    }

    // “X students only”
    const stu = /(.+?) students only/i.exec(part);
    if (stu) {
      req.studentGroup = stu[1].trim();
    }

    // “Not open to ... students”
    const ban = /Not open to (.+?) students/i.exec(part);
    if (ban) {
      req.bannedPlans = (req.bannedPlans || []).concat(planSplit(ban[1]));
    }

    // Inline whitelist example (extend with your other programs)
    const whitelistRe = /(?:Accounting and Financial Management|Computing and Financial Management|Mathematics\/Chartered Professional Accountancy|Biotechnology\/Chartered Professional Accountancy)/gi;
    const wit = [...(part.matchAll(whitelistRe))].map(m => m[0].trim());
    if (wit.length) {
      req.requiredPlans = (req.requiredPlans || []).concat(wit);
    }
  }
}
