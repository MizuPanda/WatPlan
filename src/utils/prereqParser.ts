// src/utils/prereqParser.ts

/** AST node types for prerequisite expressions */
export type PrereqNode =
  | { type: 'AND'; children: PrereqNode[] }
  | { type: 'OR';  children: PrereqNode[] }
  | { type: 'COURSE'; code: string }
  | { type: 'GRADE';  code: string; minPercent: number }
  | { type: 'PROGRAM'; program: string }

/**
 * Parses a string like
 *   "(CS 146 and CS 136L) or (grade of 60% or higher in CS 138)"
 * into a nested AST of PrereqNode.
 */
export function parsePrereqExpression(input: string): PrereqNode {
  const s0 = input.trim()
  if (!s0) return { type: 'AND', children: [] }
  return parseOr(s0)
}

function parseOr(s: string): PrereqNode {
  const parts = splitTopLevel(s, /^ or\s+/i)
  if (parts.length > 1) {
    return { type: 'OR', children: parts.map(p => parseAnd(p)) }
  }
  return parseAnd(s)
}

function parseAnd(s: string): PrereqNode {
  const parts = splitTopLevel(s, /^ and\s+/i)
  if (parts.length > 1) {
    return { type: 'AND', children: parts.map(p => parseTerm(p)) }
  }
  return parseTerm(s)
}

function parseTerm(s: string): PrereqNode {
  let str = s.trim()

  // Parenthesized group
  if (str.startsWith('(') && str.endsWith(')')) {
    return parseOr(str.slice(1, -1))
  }

  // GRADE nodes: "grade of 60% or higher in CS 138"
  const g = str.match(/grade of\s+(\d+)%\s+or higher in\s+([A-Za-z0-9 ]+\d+[A-Za-z]?)/i)
  if (g) {
    return { type:'GRADE', code:g[2].trim(), minPercent: parseInt(g[1],10) }
  }

  // COURSE nodes: look for e.g. "CS 246", "STAT 202"
  const c = str.match(/([A-Za-z0-9\/ ]+\d+[A-Za-z]?)/)
  if (c) {
    return { type:'COURSE', code: c[1].trim() }
  }

  // PROGRAM nodes: "Honours Mathematics students only"
  const p = str.match(/([A-Za-z &\/]+ students only)/i)
  if (p) {
    return { type:'PROGRAM', program: p[1].trim() }
  }

  // Fallback: treat entire segment as a PROGRAM restriction
  return { type:'PROGRAM', program: str }
}

/**
 * Splits `s` on `sep` regex only at depthâ€‘0 (i.e. not inside parentheses).
 * `sep` must be anchored (e.g. /^ or\s+/i or /^ and\s+/i).
 */
function splitTopLevel(s: string, sep: RegExp): string[] {
  const parts: string[] = []
  let depth = 0, last = 0
  for (let i = 0; i < s.length; i++) {
    if (s[i] === '(') depth++
    else if (s[i] === ')') depth--
    else if (depth === 0) {
      const substr = s.slice(i)
      const m = sep.exec(substr)
      if (m && m.index === 0) {
        parts.push(s.slice(last, i))
        i += m[0].length - 1
        last = i + 1
      }
    }
  }
  parts.push(s.slice(last))
  return parts.map(p => p.trim()).filter(p => p.length > 0)
}
