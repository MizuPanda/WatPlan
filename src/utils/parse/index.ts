// /src/parse/index.ts

import { parseExpression }          from './exprParser';
import { scanRestrictions }         from './restrictionScanner';
import { Requirements, CourseRef, AST } from './types';

/** Word→number for “One of…”, etc. */
const WORD2NUM: Record<string, number> = {
  one:1, two:2, three:3, four:4, five:5,
  six:6, seven:7, eight:8, nine:9
};

/**
 * Turn a comma/semicolon/slash list into CourseRef[].
 * Expands "SDS/SWREN250R" → SDS250R + SWREN250R,
 *         "SOC/LS280"       → SOC280   + LS280.
 */
function parseCourseRefs(text: string): CourseRef[] {
  const courses: CourseRef[] = [];
  let lastSubj: string | null = null;

  // collapse "SUBJ 123" → "SUBJ123"
  const norm = text.replace(/([A-Za-z]+)\s+(\d{3}[A-Za-z]?)/g, '$1$2');

  // split on commas or semis
  const chunks = norm.split(/[,;]/).map(s => s.trim()).filter(Boolean);

  for (const chunk of chunks) {
    // split on slash
    const parts = chunk.split('/').map(s => s.trim()).filter(Boolean);

    // collect numeric parts and letter parts
    type NumPart = { subj: string | null; cat: string };
    const numParts: NumPart[] = [];
    const letterParts: string[] = [];

    for (const p of parts) {
      const up = p.toUpperCase();
      const m = /^([A-Z]+)?(\d{3}[A-Z]?)$/.exec(up);
      if (m) {
        // m[1] might be undefined if p is just "250R"
        const subj = m[1] ?? null;
        numParts.push({ subj, cat: m[2] });
      } else if (/^[A-Z]+$/.test(up)) {
        letterParts.push(up);
      }
    }

    // if we saw any numeric parts, expand them
    if (numParts.length > 0) {
      for (const np of numParts) {
        // generate for the part itself if it had a subject
        if (np.subj) {
          courses.push({ subject: np.subj, catalog: np.cat });
          lastSubj = np.subj;
        }
        // also generate for each letter-only part
        for (const lp of letterParts) {
          courses.push({ subject: lp, catalog: np.cat });
          // we could set lastSubj = lp, but it's not needed for antireq
        }
      }

    // no numeric parts, but letterParts exist: attach to lastSubj
    } else if (letterParts.length > 0 && lastSubj) {
      for (const lp of letterParts) {
        courses.push({ subject: lp, catalog: lastSubj });
      }
    }
  }

  return courses;
}

export function parseRequirementsString(desc: string | null): Requirements {
  if (!desc) return {};
  const req: Requirements = {};

  const segments = desc.split(/(?=(?:Prereq|Coreq|Antireq):)/i);
  for (const seg of segments) {
    const m = /^((?:Prereq|Coreq|Antireq)):\s*(.+)$/i.exec(seg.trim());
    if (!m) {
      // trailing restrictions
      scanRestrictions(seg, req);
      continue;
    }

    const tag  = m[1].toLowerCase();
    const body = m[2].trim();

    // ─── Prereq ───
    if (tag === 'prereq') {
      // pure "XYZ students only"
      if (/^[A-Za-z/&\s-]+students only\.?$/i.test(body)) {
        req.studentGroup = body.replace(/students only\.?/i, '').trim();
        continue;
      }

      // One of / Two of groups
      const nof = /^(One|Two|Three|Four|Five|Six|Seven|Eight|Nine) of\s+(.+?)(?:[.;]|$)/i.exec(body);
      if (nof) {
        const count = WORD2NUM[nof[1].toLowerCase()] || 1;
        const listText = nof[2];
        const nodes = parseCourseRefs(listText);
        req.prereq = { op: 'N_OF', n: count, nodes };

        // scan any restrictions after the first period/semicolon
        const rest = body.slice(nof[0].length);
        if (rest.trim()) scanRestrictions(rest, req);
        continue;
      }

      // fallback: generic expr + restrictions
      const [exprPart, ...rest] = body.split(';');
      const ast = parseExpression(exprPart.trim());
      if (ast) req.prereq = ast;
      if (rest.length) scanRestrictions(rest.join(';'), req);

    // ─── Coreq ───
    } else if (tag === 'coreq') {
      if (/^[A-Za-z/&\s-]+students only\.?$/i.test(body)) {
        req.studentGroup = body.replace(/students only\.?/i, '').trim();
        continue;
      }
      const [exprPart, ...rest] = body.split(';');
      const ast = parseExpression(exprPart.trim());
      if (ast) req.coreq = ast;
      if (rest.length) scanRestrictions(rest.join(';'), req);

    // ─── Antireq ───
    } else if (tag === 'antireq') {
      const courses = parseCourseRefs(body);
      if (courses.length) req.antireq = courses;
    }
  }

  return req;
}
