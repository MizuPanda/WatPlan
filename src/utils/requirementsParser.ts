// src/utils/requirementsParser.ts

import { parsePrereqExpression, PrereqNode } from './prereqParser';

/** Represents a COURSE prereq (no minimum grade) */
export interface PrereqItem {
  code: string;
  name: string;
  group: number;
  relation: 'AND' | 'OR';
}

/** Represents a grade‑based prereq (e.g. “60% in CS 138”) */
export interface GradeReqItem {
  code: string;
  name: string;
  minPercent: number;
  group: number;
}

/** Represents a coreq (must take alongside) */
export interface CoreqItem {
  code: string;
  name: string;
}

/** Represents an antireq (cannot take alongside) */
export interface AntireqItem {
  code: string;
  name: string;
}

/** Represents a program restriction (e.g. “Math students only”) */
export interface ProgramRestriction {
  programCode: string;
  restriction: string;
}

/** Represents a level restriction (e.g. “Level ≥ 2A”) */
export interface LevelRestriction {
  comparator: string;
  level: string;
}

/** Represents a unit‑based prereq (e.g. “0.50 unit in HIST”) */
export interface UnitRequirement {
  subject: string;
  minUnits: number;
}

/** Aggregates all requirement types for one course */
export interface Requirements {
  prereqs: PrereqItem[];
  gradeReqs: GradeReqItem[];
  coreqs: CoreqItem[];
  antireqs: AntireqItem[];
  programRestrictions: ProgramRestriction[];
  levelRestrictions: LevelRestriction[];
  unitRequirements: UnitRequirement[];
}

/**
 * Parses a full requirementsDescription string into structured arrays.
 */
export function parseRequirements(input: string): Requirements {
  const raw = input.trim();

  const prereqMatch  = raw.match(/Prereq:\s*([^;]+)/i);
  const coreqMatch   = raw.match(/Coreq:\s*([^;]+)/i);
  const antireqMatch = raw.match(/Antireq:\s*([^;]+)/i);

  const prereqText  = prereqMatch  ? prereqMatch[1].trim() : '';
  const coreqText   = coreqMatch   ? coreqMatch[1].trim() : '';
  const antireqText = antireqMatch ? antireqMatch[1].trim() : '';

  // Strip out those clauses to leave “rest” for program/level/unit
  let rest = raw
    .replace(prereqMatch?.[0]  ?? '', '')
    .replace(coreqMatch?.[0]   ?? '', '')
    .replace(antireqMatch?.[0] ?? '', '')
    .replace(/;+/g, ';')
    .trim();

  // PROGRAM restrictions
  const programRestrictions: ProgramRestriction[] = [];
  for (const m of rest.matchAll(/([A-Za-z &\/]+) students only/gi)) {
    programRestrictions.push({
      programCode: m[1].trim(),
      restriction: 'Students Only'
    });
  }

  // LEVEL restrictions
  const levelRestrictions: LevelRestriction[] = [];
  for (const m of rest.matchAll(/Level at least\s+(\d+[AB])/gi)) {
    levelRestrictions.push({
      comparator: '>=',
      level: m[1]
    });
  }

  // UNIT requirements
  const unitRequirements: UnitRequirement[] = [];
  for (const m of rest.matchAll(/At least\s+([0-9.]+)\s+unit in\s+([A-Za-z]+)/gi)) {
    unitRequirements.push({
      subject: m[2].trim(),
      minUnits: parseFloat(m[1])
    });
  }

  // PREREQ AST → flat arrays
  const prereqs: PrereqItem[] = [];
  const gradeReqs: GradeReqItem[] = [];
  if (prereqText) {
    const ast = parsePrereqExpression(prereqText);
    flattenPrereqAST(ast, prereqs, gradeReqs);
  }

  // COREQs & ANTIREQs via simple split
  const coreqs: CoreqItem[] = coreqText
    ? splitCodes(coreqText)
    : [];
  const antireqs: AntireqItem[] = antireqText
    ? splitCodes(antireqText)
    : [];

  return {
    prereqs,
    gradeReqs,
    coreqs,
    antireqs,
    programRestrictions,
    levelRestrictions,
    unitRequirements
  };
}

/**
 * Recursively walks the PrereqNode AST and populates the flat arrays.
 * OR‑nodes share one groupId; AND‑nodes increment groupId per child.
 */
function flattenPrereqAST(
  node: PrereqNode,
  prereqs: PrereqItem[],
  gradeReqs: GradeReqItem[],
  groupId = 1
): number {
  if (node.type === 'AND') {
    for (const child of node.children) {
      groupId = flattenPrereqAST(child, prereqs, gradeReqs, groupId);
      groupId++;
    }
    return groupId;
  } else if (node.type === 'OR') {
    for (const child of node.children) {
      flattenPrereqAST(child, prereqs, gradeReqs, groupId);
    }
    return groupId;
  } else if (node.type === 'COURSE') {
    prereqs.push({
      code: node.code,
      name: node.code,
      group: groupId,
      relation: 'OR'
    });
    return groupId;
  } else if (node.type === 'GRADE') {
    gradeReqs.push({
      code: node.code,
      name: node.code,
      minPercent: node.minPercent,
      group: groupId
    });
    return groupId;
  }
  return groupId;
}

/** Splits comma/or lists of codes into {code,name} pairs */
function splitCodes(text: string): { code: string; name: string }[] {
  return text
    .split(/,|\bor\b/gi)
    .map(s => s.trim())
    .filter(s => s.length > 0)
    .map(code => ({ code, name: code }));
}
