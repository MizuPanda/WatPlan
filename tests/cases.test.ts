import { parseRequirementsString } from '../src/utils/parse';

import examples from './cases.json';

describe('cases.json', ()=>{
    for(const [label, desc] of Object.entries(examples)){
      test(label, ()=>{
        const parsed = parseRequirementsString(desc);
        expect(parsed).toMatchSnapshot();
      });
    }
  });