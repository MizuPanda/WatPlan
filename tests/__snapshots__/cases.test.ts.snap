// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cases.json CO1 1`] = `
{
  "minLevel": {
    "term": "A",
    "year": 2,
  },
}
`;

exports[`cases.json CS1 1`] = `
{
  "coreq": {
    "n": undefined,
    "nodes": [
      {
        "catalog": "136",
        "subject": "CS",
      },
      {
        "catalog": "146",
        "subject": "CS",
      },
    ],
    "op": "OR",
  },
  "prereq": {
    "n": 1,
    "nodes": [
      {
        "catalog": "145",
        "subject": "CS",
      },
      {
        "catalog": "115",
        "minGrade": 90,
        "subject": "CS",
      },
      {
        "catalog": "116",
        "minGrade": 70,
        "subject": "CS",
      },
      {
        "catalog": "135",
        "minGrade": 60,
        "subject": "CS",
      },
    ],
    "op": "N_OF",
  },
}
`;

exports[`cases.json CS2 1`] = `
{
  "minLevel": {
    "term": "A",
    "year": 3,
  },
  "prereq": {
    "n": undefined,
    "nodes": [
      {
        "catalog": "239",
        "subject": "MATH",
      },
      {
        "catalog": "249",
        "subject": "MATH",
      },
    ],
    "op": "OR",
  },
}
`;

exports[`cases.json CS3 1`] = `
{
  "antireq": [
    {
      "catalog": "292",
      "subject": "BME",
    },
    {
      "catalog": "393",
      "subject": "BME",
    },
    {
      "catalog": "241",
      "subject": "CS",
    },
    {
      "catalog": "241E",
      "subject": "CS",
    },
    {
      "catalog": "251",
      "subject": "CS",
    },
    {
      "catalog": "251E",
      "subject": "CS",
    },
    {
      "catalog": "192",
      "subject": "SYDE",
    },
  ],
  "bannedPlans": [
    "Honours Computer Science",
    "Honours Data Science (BCS",
    "BMATH)",
    "BCFM",
    "BSE",
  ],
  "prereq": {
    "n": 1,
    "nodes": [
      {
        "catalog": "116",
        "subject": "CS",
      },
      {
        "catalog": "136",
        "subject": "CS",
      },
      {
        "catalog": "138",
        "subject": "CS",
      },
    ],
    "op": "N_OF",
  },
}
`;

exports[`cases.json ECON1 1`] = `
{
  "prereq": {
    "nodes": [
      {
        "catalog": "322",
        "subject": "ECON",
      },
    ],
    "op": "AND",
  },
}
`;

exports[`cases.json FINE1 1`] = `
{
  "milestones": [
    "Fine Arts Health and Safety Milestone",
    "Fine Arts Woodshop Workshop Milestone",
  ],
  "prereq": {
    "nodes": [
      {
        "catalog": "100",
        "subject": "FINE",
      },
    ],
    "op": "AND",
  },
}
`;

exports[`cases.json GER1 1`] = `
{
  "prereq": {
    "n": 1,
    "nodes": [
      {
        "catalog": "201",
        "subject": "GER",
      },
      {
        "catalog": "211",
        "subject": "GER",
      },
    ],
    "op": "N_OF",
  },
}
`;

exports[`cases.json GLOB1 1`] = `
{
  "studentGroup": "Global Governance Doctoral",
}
`;

exports[`cases.json HIST1 1`] = `
{
  "minLevel": {
    "term": "A",
    "year": 2,
  },
  "minUnits": 0.5,
  "minUnitsSubject": "HIST",
}
`;

exports[`cases.json LVL1 1`] = `
{
  "minLevel": {
    "term": "A",
    "year": 3,
  },
}
`;

exports[`cases.json OPTOM1 1`] = `
{
  "prereq": {
    "nodes": [
      {
        "catalog": "106",
        "subject": "OPTOM",
      },
    ],
    "op": "AND",
  },
  "studentGroup": "Optometry",
}
`;

exports[`cases.json PACS1 1`] = `
{
  "prereq": {
    "nodes": [
      {
        "n": undefined,
        "nodes": [
          {
            "catalog": "202",
            "subject": "PACS",
          },
          {
            "catalog": "271",
            "subject": "LS",
          },
        ],
        "op": "OR",
      },
      {
        "n": undefined,
        "nodes": [
          undefined,
          {
            "n": undefined,
            "nodes": [
              undefined,
              undefined,
            ],
            "op": "AND",
          },
        ],
        "op": "OR",
      },
    ],
    "op": "AND",
  },
}
`;

exports[`cases.json PHAR1 1`] = `
{
  "studentGroup": "Pharmacy",
}
`;

exports[`cases.json PHARM2 1`] = `{}`;

exports[`cases.json PSYCH1 1`] = `
{
  "prereq": {
    "n": undefined,
    "nodes": [
      {
        "catalog": "101",
        "subject": "LS",
      },
      {
        "catalog": "206",
        "subject": "LS",
      },
      {
        "catalog": "260",
        "subject": "PSCI",
      },
    ],
    "op": "OR",
  },
}
`;

exports[`cases.json SOC1 1`] = `
{
  "prereq": {
    "n": undefined,
    "nodes": [
      {
        "catalog": "101",
        "subject": "SOC",
      },
      {
        "catalog": "101R",
        "subject": "SOC",
      },
    ],
    "op": "OR",
  },
}
`;

exports[`cases.json SOC2 1`] = `
{
  "minLevel": {
    "term": "A",
    "year": 3,
  },
  "prereq": {
    "n": undefined,
    "nodes": [
      {
        "catalog": "101",
        "subject": "SOC",
      },
      {
        "catalog": "101R",
        "subject": "SOC",
      },
    ],
    "op": "OR",
  },
}
`;
